internal/pubsub/gcppubsub/gcppubsub.go

```
@@ -0,0 +1,130 @@
+}
+
+// Close closes the connection to the API service. The user should invoke this
+// when the client is no longer required.
+func (t *topic) Close() error {
+	return t.client.Close()
+}
+
+// SendBatch publishes all the messages in dms. This method returns only after
+// all the messages are sent, an error occurs, or the context is cancelled.
+func (t *topic) SendBatch(ctx context.Context, dms []*driver.Message) error {
+	var ms []*pubsubpb.PubsubMessage
+	for _, dm := range dms {
+		m := &pubsubpb.PubsubMessage{
+			Data:       dm.Body,
+			Attributes: dm.Metadata,
+		}
+		ms = append(ms, m)
+	}
+	req := &pubsubpb.PublishRequest{
+		Topic:    t.path,
+		Messages: ms,
+	}
+	_, err := t.client.Publish(ctx, req)
+	return err
+}
+
+type subscription struct {
+	client *rawgcppubsub.SubscriberClient
+	path   string
+}
+
+// ReceiveBatch returns a batch of messages that have queued up for the
+// subscription on the server. If no messages are available yet, it blocks
+// until there is at least one, or the context is done.
+func (s *subscription) ReceiveBatch(ctx context.Context) ([]*driver.Message, error) {
+	req := &pubsubpb.PullRequest{
+		Subscription:      s.path,
+		ReturnImmediately: false,
+	}
+	resp, err := s.client.Pull(ctx, req)
+	var ms []*driver.Message
+	for _, rm := range resp.ReceivedMessages {
+		rmm := rm.Message
+		m := &driver.Message{
+			Body:     rmm.Data,
+			Metadata: rmm.Attributes,
+			AckID:    rm.AckId,
+		}
+		ms = append(ms, m)
+	}
+	return ms, err
+}
+
+// SendAcks acknowledges the messages with the given ackIDs on the server so
+// that they will not be received again for this subscription if the server
+// gets the acks before their deadlines. This method returns only after all
+// the ackIDs are sent, an error occurs, or the context is cancelled.
+func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
+	var ids2 []string
+	for _, id := range ids {
+		id2, ok := id.(string)
+		if !ok {
```

Meh, I'd just let it panic.

---

Done.